homeassistant/components/recorder/core.py

```diff
@@ -681,22 +686,91 @@ def run(self) -> None:
-            try:
-                self._process_one_task_or_recover(task)
-            except Exception as err:  # pylint: disable=broad-except
-                _LOGGER.exception("Error while processing event %s: %s", task, err)
-
-        self._shutdown()
+            self._guarded_process_one_task_or_recover(queue_.get())
+
+    def _pre_process_startup_tasks(self, startup_tasks: list[RecorderTask]) -> None:
+        """Pre process startup tasks."""
+        # Prime all the state_attributes and event_data caches
+        # before we start processing events
+        state_change_events: list[Event] = []
+        non_state_change_events: list[Event] = []
+
+        for task in startup_tasks:
+            if isinstance(task, EventTask):
+                event_ = task.event
+                if event_.event_type == EVENT_STATE_CHANGED:
+                    state_change_events.append(event_)
+                else:
+                    non_state_change_events.append(event_)
+
+        self._pre_process_state_change_events(state_change_events)
+        self._pre_process_non_state_change_events(non_state_change_events)
+
+    def _pre_process_state_change_events(self, events: list[Event]) -> None:
+        """Load startup state attributes from the database.
+
+        Since the _state_attributes_ids cache is empty at startup
+        we restore it from the database to avoid having to look up
+        the attributes in the database for every state change
+        until its primed.
+        """
+        assert self.event_session is not None
+        if hashes := [
+            StateAttributes.hash_shared_attrs_bytes(shared_attrs_bytes)
+            for event in events
+            if (
+                shared_attrs_bytes := self._serialize_state_attributes_from_event(event)
+            )
+        ]:
+            with self.event_session.no_autoflush:
+                for id_, shared_attrs in self.event_session.execute(
+                    get_shared_attributes(hashes)
+                ).fetchall():
+                    self._state_attributes_ids[shared_attrs] = id_
+
+    def _pre_process_non_state_change_events(self, events: list[Event]) -> None:
+        """Load startup event attributes from the database.
+
+        Since the _event_data_ids cache is empty at startup
+        we restore it from the database to avoid having to look up
+        the data in the database for every event until its primed.
+        """
+        assert self.event_session is not None
+        if hashes := [
+            EventData.hash_shared_data_bytes(shared_event_bytes)
+            for event in events
+            if (shared_event_bytes := self._serialize_event_data_from_event(event))
+        ]:
+            with self.event_session.no_autoflush:
+                for id_, shared_data in self.event_session.execute(
+                    get_shared_event_datas(hashes)
+                ).fetchall():
+                    self._event_data_ids[shared_data] = id_
```

Comment:
I need to batch these in groups of 999 since we still support ancient versions of sqlite
[SQLite](https://www.sqlite.org/) ≥ 3.31.0

Reply:
6635a9144ca4abbdbe9525a970bbc2a3d0b5aa39